src/Microsoft.Extensions.Configuration.AzureAppConfiguration/AzureAppConfigurationProvider.cs

```diff
@@ -450,6 +468,18 @@ private async Task RefreshIndividualKeyValues()
             }
         }
 
+        private async Task RefreshKeyValueAdapters()
+        {
+            foreach (IKeyValueAdapter adapter in _options.Adapters)
+            {
+                if (adapter.NeedsRefresh())

```

Comment:
>adapter.NeedsRefresh() [](http://example.com/codeflow?start=20&length=22)

If `RefreshAsync` is called many times rapidly we don't want to call SetData each time if the initial refresh operation has started. This is a real scenario for high throughput web apps using our middleware. It looks like we need a semaphore that is tied to the adapter.

Reply:
added a semaphore to fix this issue.